### ISSUE TITLE
discovery of dune - questions

### ISSUE THREAD

**Issue by MdeLv** [ocaml/dune#3429]
### 1/ hello_world example
Trying to run the simple hello_world example, I got nothing with `dune build @install` and `dune runtest`. I had to look up in the the documentation, then in the issues where I found [running dune in hello_world project does nothing #1584](https://github.com/ocaml/dune/issues/1584).
I discovered that a `--root .` option/parameter are needed in order to override the "silent outermost dune workspace file supremacy".
That's pretty deceptive for a first contact, and it lay be even more deceptive for s.o. who is also new to OCaml.
What about simply adding the following to README.org?

> To build sample-projects/hello_world FROM this directory, you MUST specify:
> $ dune build --root .  
> (because dune chooses the outermost workspace dune file, unless specified differently).
> 
> To clean up your build directory, do:
> $ dune clean --root .
> 
> To see dune's actions, for instance, add the `--verbose` option, as in:
> $ dune build --root . --verbose

### 2/ dry-run ?
The option `--verbose` is needed to understand what is going on under dune's hood.
With the existing options, is it possible to do a dry run that would just display the actions intended by dune?

###  3/ CLI autocompletion (bash complete)
Do you have a shell script for dune's commands completion? Such a _complete_ shell script, available for opam, saves time and helps to learn the available commands **and** options.

###  4/ dune's syntax
Do you have a BNF-like definition of dune's syntax? I personally prefer read such clear definition instead of digging into a [semi-formal textual documentation](https://dune.readthedocs.io).

###  5/ checking dune files globally
The lint tool helps to check the syntax of one dune file (btw, is there one for dune files, or is there just a way to use a lint tool for OCaml programs?)
There should quickly be a bunch of dune files spread throughout all the files tree/workspace. And I suppose that handling so many dune files required for a large program can be problematic. 
Is there a way to use existing dune tools in order to check that all these dune files make a coherent set of build directives?

**Comment by jeremiedimino** [comment]
### 1/ hello_world example

The example directory predates [the quickstart](https://dune.readthedocs.io/en/stable/quick-start.html). We should probably just remove it to avoid having to maintain two things in parallel for the same purpose. Did you  have a look at the quick start BTW?

### 2/ dry-run ?

It's difficult for dune to print such action without doing anything because computing some of the command lines requires running actions. The closest that exist is `dune rules -m <target>`, which prints a makefile to build `<target>`.

### 3/ CLI autocompletion (bash complete)

Dune uses the [cmdliner library](https://github.com/dbuenzli/cmdliner) to implement the CLI and auto-completion is pending on [this issue](https://github.com/dbuenzli/cmdliner/issues/1).

### 4/ dune's syntax

That's coming :) Maintaining a BNF definition manually on the side would be too much work, but I'm doing some work on our parsing engine. Among other things, this work will make it possible to extract a BNF definition from the parser itself.

### 5/ checking dune files globally

Dune systematically reads all the dune files and performs various checks. So a lint tool feels a little superfluous. But I suppose it wouldn't hurt to add a `lint` command that stops right after reading the dune files.

**Comment by MdeLv** [comment]
> 1/ The example directory predates [the quickstart](https://dune.readthedocs.io/en/stable/quick-start.html). We should probably just remove it to avoid having to maintain two things in parallel for the same purpose. Did you have a look at the quick start BTW?

Yes, I started with quick start.
I played with the example/ directory because I saw a clear reference to it in [README.md](https://github.com/ocaml/dune/blob/master/README.md): "The [example](https://github.com/ocaml/dune/tree/master/example) directory contains examples of projects using dune".
Maybe this example could be introduced at the end of the quick start to make people aware of what happens when they play within a set of directories. It's a chance to introduce the `--root` option in quick start where it's totally absent.

>2/  `dune rules -m <target>`, which prints a makefile to build `<target>`.

The commands that work for me are:
`dune rules -m` (outputs a bunch of text! )
`dune rules <path-to-dune-file> -m`

`dune rules -m <target>` with target=Makefile or Foo produces an error:
Error: Don't know how to build <target>
(dune.2.5.1)

It appears that what you call a `<target>` is the argument of the `dune rules` command, while for me it's the file outputed by this command.

Then, to get a Makefile, I had to do:
`dune rules <path-to-dune-file> -m > Makefile`
because
`dune rules <path-to-dune-file> -o Makefile` causes an error:
```
Internal error, please report upstream including the contents of _build/log.
Description:
  ("Fdecl.get: not set", {})
Description:
  ("Fdecl.get: not set", {})
Raised at file "src/stdune/code_error.ml", line 9, characters 30-62
```
Is the syntax `-o <outputed-file>` correct?
It was not very smooth, but at last it works! 

Using a dune file  as an argument produces a readable Makefile of a few lines, which is fine.
I get hundredths of lines when no argument is given (dump all the internal rules.)
Can you give a few words about all that stuff? (optional open question)

> 3/ Dune uses the [cmdliner library](https://github.com/dbuenzli/cmdliner) to implement the CLI and auto-completion is pending on [this issue](https://github.com/dbuenzli/cmdliner/issues/1).

Can you give a time frame, and a summary (what has be done/what needs to be done)?

> 4/ Maintaining a BNF definition manually on the side would be too much work, but I'm doing some work on our parsing engine.

I have a very quick look at the source code. Can you pls. give a pointer to the implementation of dune concrete syntax parsing in [dune/tree/master/src](https://github.com/ocaml/dune/tree/master/src)?

> **5/** checking dune files globally

Yes, dune must do the job globally within a dune-workspace .
A `dune lint` command would be nice to check individual files and learn. In between, can you give a reference to a lint function within the source code?

Thanks.

**Comment by jeremiedimino** [comment]
The `-o` behaviour does look like a bug. Do you mind opening a bug report for this?

Regarding the general behaviour of `dune rules` or interpretation of `<target>`, it seems to me that the documentation given by `dune rules --help` is clear enough.

> Can you give a time frame, and a summary (what has be done/what needs to be done)?

Someone needs to implement the feature in cmdliner, and then it should immediately become available in dune. That's not a priority at the moment and no one from the dune team is planning to work on it in the short/medium term.

> I have a very quick look at the source code. Can you pls. give a pointer to the implementation of dune concrete syntax parsing in dune/tree/master/src?

Sure. We are using a parsing monad defined in `src/dune_lang/decoder.mli`. The parser for dune files is spread across several files in the code base, but most of it can be found in `src/dune/dune_file.ml`. Look for things such as `let+` or `field "` in this file to find the functions that parse various parts of the dune files.

I'm planning to turn this monad into an applicative, which will allows us to extract the BNF directly from a `Decoder.t`.

> In between, can you give a reference to a lint function within the source code?

There is no such function, the checks are done as part of the parsing. You can search for `User_error.raise` in `src/dune/dune_file.ml` to find several such checks.

**Comment by MdeLv** [comment]
> The `-o` behaviour does look like a bug. Do you mind opening a bug report for this?

Done: #3440.

> it seems to me that the documentation given by `dune rules --help` is clear enough

For me, not immediately. In the context of make, a _target_ is the file that is outputted, which was confusing. 
In the context of a dune file, I see that `<target>` is also the name of the outputted file, as confirmed by quick start:  [Using a custom code generator](https://dune.readthedocs.io/en/stable/quick-start.html#using-a-custom-code-generator)

`man dune rules` tells that  _target_ is used on the same page with different meanings (`TARGET` for the dune file and `target` for the outputted file):

> SYNOPSIS
>        dune rules [OPTION]... [TARGET]...
>  DESCRIPTION
>        By default the output is a list of S-expressions, one S-expression per
>        rule. Each S-expression is of the form:
>          ((deps    (<dependencies>))
>           (targets (<targets>))
>           (context <context-name>)
>           (action  <action>))
> 
BTW, it's a good idea to include a Makefile so that people can invoke dune with the well known `make && make install` (at first sight, a dune invocation may be seen as too subtle because one needs to know the way a dune workspace is managed by dune).

I thank  for your references to the dune parser.

> I'm planning to turn this monad into an applicative, which will allows us to extract the BNF directly from a `Decoder.t`.

Great. When do you expect to do complete that?

**Comment by jeremiedimino** [comment]
> For me, not immediately. In the context of make, a target is the file that is outputted, which was confusing.
> In the context of a dune file, I see that <target> is also the name of the outputted file, as confirmed by quick start: Using a custom code generator

That's a good point. Maybe we could call this `OUTPUT-FILE`. Could you send a PR to change the doc? It's in the file `bin/print_rules.ml`.

> BTW, it's a good idea to include a Makefile so that people can invoke dune with the well known make && make install (at first sight, a dune invocation may be seen as too subtle because one needs to know the way a dune workspace is managed by dune).

Totally agreed, and that's why we do include one in dune :) I also include one in all my projects.

> Great. When do you expect to do complete that?

Not sure, it's not a priority at this point. I'm just working on it when I can find some time because I find the subject interesting.

**Comment by MdeLv** [comment]
> That's a good point. Maybe we could call this `OUTPUT-FILE`. Could you send a PR to change the doc? It's in the file `bin/print_rules.ml`.



`dure rules --helps` says:

SYNOPSIS
       dune rules [OPTION]... [TARGET]...
...
OPTIONS

       -o FILE
           Output to a file instead of stdout.
       -m, --makefile
           Output the rules in Makefile syntax.

Normal behavior of `dune rules` is outputting to stdout.
What does [TARGET] mean? The optional dune file we want to target, or really the file we want to output? 
I tried to target a dune file in sub directory but got the same result.

So, I think that  synopsis should be simply:
       dune rules [OPTION]...
or (redundant with options):
       dune rules [OPTION]... [-o OUTPUT]...
Ok?

> ### 2/ dry-run ?
> 
> It's difficult for dune to print such action without doing anything because computing some of the command lines requires running actions. The closest that exist is `dune rules -m <target>`, which prints a makefile to build `<target>`.

What about doing all the actions and outputting the files in a sandbox, located somewhere in _build/default/sandbox directory that would be created and immediately removed. 
On linux, it could be simply outputting generated files to /dev/null .
What do you think about that?

### Concerning this "dune discovery":
I'm convinced that the core of dune is robust, although it's not formal. But the user experience is somehow painful.
From my experience, this is mainly because: dune does things without telling what it does. Or does nothing without telling it, and why.  Or does not do  things the user expects it should do.
As a consequence, the user doesn't really know if he is in a safe place or near a trap. He has to dig again an again until he can predict what will happen.
You can no more be aware of that because you develop and use dune intensively.

The [quickstart](https://dune.readthedocs.io/en/stable/quick-start.html#) seems attractive at first sight but is confusing.
Doing `$ dune build hello_world.exe` is fine.
`$ ./_build/default/hello_world.exe` starts to be too heavy.
Why not making a symlink in the workspace root directory? It's visible: I see that something was produced without digging into _build/ and I can simply launch it with a basic `./hello_world.exe` command.

I tried to make a symlink to exe in [Defining tests](https://dune.readthedocs.io/en/stable/quick-start.html#defining-tests) from quickstart. Maybe I didn't well configured this very simple files tree, but I had the following message:
```
$ dune build
Error: Multiple rules generated for _build/default/main.exe:
- file present in source tree
- dune:2
Hint: rm -f main.exe
```
On that very simple I don't know what to do except removing the symlink.
I didn't have this issue with the example hello_world herebefore.

How can we  protect this symlink created or declare to dune "add a symlink and preserve it" ?

opam file: 
it's nice to tell the user that the package must declare a side-library (of the program) in opam, but why not proposing to add it (at field `depends`)?

Once I've created a set of signature files that are my program ready to implement, declared the libraries it uses (lwt, etc.) and declared the side-library created, there is all the information for an opam file (I think).
I would like this opam file to be generated or updated by dune, according to the program's dependencies and developed library.
How can this be configured or implemented?

And if If I have the structure files, there is even all the information about the libraries to feed the dune files. It could even be done incrementally.
Which dune primitives can be used to build the dune files from the .ml files?

Thanks.

PS: I have other remarks concerning the naming of some commands that lacks of simplicity (compared with other mainstream PLs). We'll discuss later about that if you want.

**Comment by jeremiedimino** [comment]
This seems mostly related to the documentation and cli. @shonfeder and/or @emillon could you see if there is something worth changing in Dune?

**Comment by shonfeder** [comment]
Hi, sorry for the delay. I'm happy to do a detailed read of the exchange here and file issues/PRs to make documentation/CLI improvements as necessary, but I won't have the bandwidth until next weekend, at the earliest.

**Comment by MdeLv** [comment]
There is no urgency. I have a workaround.
Next week is fine.

btw, do you mainly use dune with CLI or programatically?

**Comment by shonfeder** [comment]
Hi @MdeLv. I use dune both ways :)

**Comment by ejgallego** [comment]
Ping @shonfeder

**Comment by shonfeder** [comment]
Hi! Sorry for the long delay here.

There's a lot of back and forth on this issue. Hard for me to figure out what it's actually about at this point!

Addressing some remarks in the the last comment from @MdeLv:

### What does [TARGET] mean?

>  What does [TARGET] mean? The optional dune file we want to target, or really the file we want to output?
> I tried to target a dune file in sub directory but got the same result.

`dune rules --help` currently says: 

       Dump Dune internal rules for the given targets. If no targets are
       given, dump all the internal rules.

This is pretty clear and consistent with the use of "target" elsewhere in the docs, e.g., in the description for `build`. It is the build target. Lots of things work on build targets, but only the build generates these targets. If someone thinks this is unclear and an idea on how to clarify this somehow, PRs to the docs would surely be welcome.

### dry-run

I don't fully understand what's being asked for around this, or how the proposed solution works address the problem explained by Jeremie. If there's still an urge around this, I suggest opening a separate issue clearly articulating what is needed (perhaps also a proposal for how that need could be met).

### re: Location of executables

>  Why not making a symlink in the workspace root directory? It's visible: I see that something was produced without digging into 
> _build/ and I can simply launch it with a basic ./hello_world.exe command.

IMO, we generally don't want dune creating any files outside of the `_build` directory (unless we've explicitly told it to because we're generating some dependency). All of the modern build systems I use on a daily basis follow this kind of convention for their Executables. In any case, dune has `dune exec hello_world`, making the location of the built executable irrelevant for the purposes of testing. This is all documented in the quick start.

### re: symlinks

I just tried creating a symlink to a built executable and it didn't produce the same error. Perhaps that's been fixed in the intervening months? If you can come up with a minimal reproduction of the error, please open an issue.

### re: opam file

I think the generation of opam files documented at https://dune.readthedocs.io/en/stable/opam.html?highlight=opam#generating-opam-files address this.


--- 

@ejgallego Thanks for the reminder here. Was there any else in particular you'd wanted addressed here?

**Comment by shonfeder** [comment]
I think we can consider this closed by virtue of the subsequent development, and the answers given there. There's instruction for opening more focused followup issues, in case that's need.
